src/guide/structs.md

```diff
@@ -0,0 +1,62 @@
+# Struct errors
+
+You may not always need the flexibility of an enum for your error
+type. In those cases, you can use the familiar SNAFU attributes with a
+struct:
+
+```rust
+# use std::convert::TryFrom;
+# use snafu::Snafu;
+#[derive(Debug, Snafu)]
+#[snafu(display("Unable to parse {} as MyEnum", value))]
+struct ParseError {
+    value: u8,
+}
+
+// That's all it takes! The rest is demonstration of how to use it.
+
+#[derive(Debug)]
+enum MyEnum {
+    Alpha,
+    Beta,
+    Gamma,
+}
+
+impl TryFrom<u8> for MyEnum {
+    type Error = ParseError;
+
+    fn try_from(other: u8) -> Result<Self, Self::Error> {
+        match other {
+            0 => Ok(Self::Alpha),
+            1 => Ok(Self::Beta),
+            2 => Ok(Self::Gamma),
+            value => ParseContext { value }.fail()
+        }
+    }
+}
+```
+
+## Differences from enum errors
+
+While each enum error variant creates a context selector that matches
+the variant's name, context selectors for structs remove the suffix
+`Error` from the name of the error, if present,  and add `Context`:
```

Comment:
It did take me a bit to understand why this renaming happened, but I think I understand now.  Probably the most direct reason is that SNAFU couldn't generate selector structs with the same name in the same scope.  Do I have that right?  If so, I'm not sure it would work to add `Snafu` to a struct that's not suffixed with `Error`.

I think the approach you have makes sense.

Out of curiosity, are there any other approaches you considered?  One that jumped to my mind, though it might be silly, is to use namespacing.  Generated selectors could be placed in a new module, either with a name given as an attribute or a default.  I'm not sure it makes sense to have multiple implementations, though, and renaming is certainly simpler, even if it causes a question or two.

Reply:
> that SNAFU couldn't generate selector structs with the same name in the same scope. Do I have that right?

exactly!

> I'm not sure it would work to add `Snafu` to a struct that's not suffixed with `Error`.

I'm not totally following what you mean here.

> Generated selectors could be placed in a new module,

This is something that's been brought up before, but I don't immediately see how it can work because procedural macros only have access to the tokens they are on. For example:

```rust
use my::thing::Foo;

#[derive(Snafu)]
enum Error {
    Alpha { foo: Foo },
}
```

We only see `Foo`, not `my::thing::Foo`, so if we generated a submodule, `Foo` would no longer be in scope:

```
mod context {
    struct Alpha { foo: Foo };
}
```

Typing that out though, the selector is actually all generic types, and we could put the implementations outside of the module:

```
use my::thing::Foo;

#[derive(Snafu)]
enum Error {
    Alpha { foo: Foo },
}

mod context {
    struct Alpha<F> { foo: F };
}

impl<F: Into<Foo>> context::Foo {
    // ...
}
```

So it might indeed be possible.

---

I see a few possible futures:

1. Change enums to *also* add `Context` to the context selector. This makes the cases uniform, and *might* reduce one of the biggest points of confusion about using SNAFU — the usage of the same name to mean two different things.
1. Add the ability to control the selector name / how it's generated anyway. I might like to prefer `Ctx` for my own crates, instead of typing out the long form.
1. Add the ability to namespace the selectors, as you mentioned. We'd have to figure out how to support multiple errors creating modules in the same parent module, as otherwise the new modules would conflict.

Reply:
>> I'm not sure it would work to add Snafu to a struct that's not suffixed with Error.

> I'm not totally following what you mean here.

I meant something very close to the first point:

> SNAFU couldn't generate selector structs with the same name in the same scope

because if you don't have a suffix of "Error", SNAFU wouldn't do any renaming, and the context selector struct would have the same name as the original struct.  It works for enums (as you say, "the usage of the same name to mean two different things") because the names of enum variants and the names of structs live in different scopes.  With this PR, the generated selector structs and the original structs (with `derive(Snafu)`) live in the same scope, so they can't have the same name, which seems to imply to me that the input structs *must* be suffixed with "Error".  Am I making a logical misstep? :)

---

> might reduce one of the biggest points of confusion about using SNAFU — the usage of the same name to mean two different things.

I did have trouble understanding that at first, and I've explained it to others, so I see where you're coming from.  I think any of the three options would solve that confusion.

I think (2) is a generalized case of (1), since it could have a default of adding "Context".  It doesn't *seem* like it'd be a lot of extra work to add the configurability, since there are somewhat similar examples already, like source transformation.

For (3), I admit I hadn't thought about multiple errors in a module, since I haven't used that yet.  It does seem like you'd want the module to be named the same for all errors, e.g. `mod context`.  With that restriction, a solution isn't jumping to mind...

I would lean toward going with what you have for now, with structs being special, and trying to think of a solution for (3)'s issues.

I worry that (1) and (2) (with adding "Context" by default) would be too much naming overhead.  I suspect that enums will still be the most common usage of SNAFU, and referencing them with something like `error::FooContext` rather than `error::Foo` would get annoying, as would (with (2)) having to specify that your desired naming transformation is to do nothing.  (Also, there's the cost of migration for existing users, but there could be ways of mitigating that, so it's not my first consideration.)

I suppose there's also the option of doing (2) just for structs, so users can tailor the special case to their liking.